cmd/keymasterd/main.go

```diff
@@ -116,9 +117,67 @@ var (
 		},
 		[]string{"username", "type"},
 	)
+	authOperationCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "keymaster_auth_operation_counter",
+			Help: "Keymaster_auth_operation_counter",
+		},
+		[]string{"client_type", "type", "result"},
+	)
+
+	externalServiceDurationTotal = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "keymaster_external_service_request_duration",
+			Help:    "Total amount of time spent non-errored external checks in ms",
+			Buckets: []float64{5, 7.5, 10, 15, 25, 50, 75, 100, 150, 250, 500, 750, 1000, 1500, 2500, 5000},
+		},
+		[]string{"service_name"},
+	)
+	tricorderLDAPExternalServiceDurationTotal    = tricorder.NewGeometricBucketer(5, 5000.0).NewCumulativeDistribution()
+	tricorderStorageExternalServiceDurationTotal = tricorder.NewGeometricBucketer(1, 2000.0).NewCumulativeDistribution()
+	tricorderVIPExternalServiceDurationTotal     = tricorder.NewGeometricBucketer(5, 5000.0).NewCumulativeDistribution()
+
+	certDurationHistogram = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "keymaster_cert_duration",
+			Help:    "Duration of certs in seconds",
+			Buckets: []float64{15, 30, 60, 120, 300, 600, 3600, 7200, 36000, 57600, 72000, 86400, 172800},
+		},
+		[]string{"cert_type", "stage"},
+	)
+
 	logger log.DebugLogger
 )
 
+func metricLogAuthOperation(clientType string, authType string, success bool) {
+	validStr := strconv.FormatBool(success)
+	metricsMutex.Lock()
+	defer metricsMutex.Unlock()
+	authOperationCounter.WithLabelValues(clientType, authType, validStr).Inc()
+}
+
+func metricLogExternalServiceDuration(service string, val float64) {
+	metricsMutex.Lock()
+	defer metricsMutex.Unlock()
+	externalServiceDurationTotal.WithLabelValues(service).Observe(val)
+	switch service {
+	case "ldap":
+		tricorderLDAPExternalServiceDurationTotal.Add(val)
```

Comment:
I notice you set these distributions to units.Millisecond below.  You can add time.Duration values instead of float64 values to such a distribution and it will do the right thing. That is it will still report the duration in milliseconds.